website/docs/language/resources/terraform-data.mdx

```diff
@@ -0,0 +1,95 @@
+---
+page_title: The terraform_data Managed Resource Type
+description: >-
+  Retrieves the root module output values from a Terraform state snapshot stored
+  in a remote backend.
+---
+
+# The `terraform_data` Resource
+
+The `terraform_data` implements the standard resource lifecycle, but does not directly take any other actions. 
+You can use the `terraform_data` resource without requiring or configuring a provider. It is always available through a built-in provider with the [source address](/language/providers/requirements#source-addresses) `terraform.io/builtin/terraform`.
+
+The `terraform_data` resource is useful for storing values which need to follow a manage resource lifecycle, and for triggering provisioners when there is no other logical managed resource in which to place them.
+
+
+## Example Usage (data for `replace_triggered_by`)
+
+```hcl
+variable "revision" {
+  default = 1
+}
+
+resource "terraform_data" "replacement" {
+  input = var.revision
+}
+
+# This resource has no convenient attribute which forces replacement,
+# but can now be replaced by any change to the revision variable value.
+resource "example_database" "test" {
+  lifecycle {
+    replace_triggered_by = [terraform_data.replacement]
+  }
+}
+```
+
+## Example Usage (`null_resource` replacement)
+
+```hcl
+resource "aws_instance" "web" {
+  # ...
+}
+
+# A use-case for terraform_data is as a do-nothing container
+# for arbitrary actions taken by a provisioner.
+resource "terraform_data" "bootstrap" {
+  triggers_replace = aws_instance.web.id
+
+  connection {
+    host = aws_instance.web.public_ip
+  }
```

Comment:
This seems like a bit of a misleading example because in this case it would be reasonable and better to put the `connection` and `provisioner` blocks directly inside the `resource "aws_instance" "web"` block.

Maybe we could make this a more "realistic" example by showing a multi-instance resource having all of its instances provisioned together somehow, like this:

```hcl
resource "aws_instance" "web" {
  count = 5

  # ...
}

resource "terraform_data" "cluster" {
  triggers_replace = aws_instance.web[*].id

  provisioner "local-exec" {
    inline = [
      "form-cluster ${join(" ", aws_instance.web[*].private_ip)}",
    ]
  }
}
```

This example could probably benefit from a leading paragraph adding some context about when this sort of approach might be needed. For example:

> [Provisioners are a last resort](https://developer.hashicorp.com/terraform/language/resources/provisioners/syntax#provisioners-are-a-last-resort), but if you need to declare one that doesn't naturally belong to only a single resource instance then you can use `terraform_data` as a generic trigger for running provisioning steps that work with multiple resource instances at the same time.

And then at the end, after the example:

> *Note:* The `connection` and `provisioner` blocks are valid inside any `resource` block. If your provisioner action involves only individual resource instances then it's simpler to declare the provisioner inline with the resource it relates to, such as inside the `resource "aws_instance" "web"` block above. In that case you do not need to use `terraform_data` at all.

(This extra note is because I've noticed that we seem to have misled some members of the community into thinking that `provisioner` is a feature of `null_resource` in particular, rather than being a meta-argument available for all resource types.)

Reply:
This example was mostly duplicated from the `Provisioners Without a Resource` section and better documented there, so I changed it slightly to reference multiple separate resources instead. In the future we should probably find a more suitable location for the info and merge the info between these two sections.